lumicks/pylake/detail/imaging_mixins.py

```diff
@@ -1,11 +1,131 @@
+import itertools
+import json
+import numpy as np
+import numpy.typing as npt
+import tifffile
+from typing import Iterator, Union
+
 from ..adjustments import ColorAdjustment
 from .timeindex import to_timestamp
-import numpy as np
-from typing import Union
 
 _FIRST_TIMESTAMP = 1388534400
 
 
+class TiffExport:
+    def export_tiff(self, filename, *, dtype=None, clip=False):
+        """Export frames to a TIFF image
+
+        Parameters
+        ----------
+        filename : str
+            The name of the TIFF file where the image will be saved.
+        dtype : np.dtype
+            The data type of a single color channel in the resulting image.
+        clip : bool
+            If enabled, the image data will be clipped to fit into the desired `dtype`. This option
+            is disabled by default: an error will be raise if the data does not fit.
+        """
+        # If the the exported tiff should be casted to `dtype`, get the full image stack to later
+        # safely cast it. Otherwise, try to get an iterator, to save memory.
+        frames = self._tiff_frames(iterator=dtype is None)
+
+        def _check_empty(frames):
+            if hasattr(frames, "__next__"):
+                # `frames` is an iterator. Peek into the first frame to check if there are any.
+                def peek(iterable):
+                    iterator = iterable
+                    try:
+                        first = next(iterator)
+                    except StopIteration:
+                        return None
+                    return itertools.chain([first], iterator)
+
+                frames = peek(frames)
+            else:
+                # `frames` is an np.ndarray. Check the size.
+                frames = frames if frames.size > 0 else None
+
+            if frames is None:
+                raise RuntimeError("Can't export TIFF if there are no pixels")
+
+            return frames
+
+        frames = _check_empty(frames)
```

Comment:
I wish there was an easier way to do this.

I was hoping we could leave this statement out entirely and just iterate; and while this works (as long as you make sure that there is a check before `_cast_image` if there are no images. I was hoping that `TiffWriter` would not write a file until a frame comes in, but unfortunately it does.

However, since we are using the `_tiff_timestamp_ranges` anyhow, it might be worth considering just checking whether that is empty. Note though that if you go this route, you should store it, because they are somewhat expensive to grab.

```python
        ts_ranges = self._tiff_timestamp_ranges()
        if not ts_ranges:
            raise RuntimeError("Can't export TIFF if there are no pixels")
```

Reply:
I originally intended to also be able to export tiffs without timestamp ranges, which is why I did not think of checking the length of `_tiff_timestamp_ranges()`. As we ensure to provide timestamp ranges this is indeed a much cleaner solution!

Implemented!